// @gag/popconfirm-web/src/index.jsx
import '../style';
import React from 'react';
import PropTypes from 'prop-types';
import Tooltip, { AbstractTooltipProps }  from '@gag/tooltip-web';
import Icon from '@gag/icon-web';
import Button from '@gag/button-web';
import injectLocale from '@gag/locale-provider-web/injectLocale';

class Popconfirm extends React.Component{

  constructor(props) {
    super(props);

    this.state = {
      visible: props.visible,
    };
  }
  componentWillReceiveProps(nextProps) {
    if ('visible' in nextProps) {
      this.setState({ visible: nextProps.visible });
    }
  }

  onConfirm = (e) => {
    this.setVisible(false);

    const { onConfirm } = this.props;
    if (onConfirm) {
      onConfirm.call(this, e);
    }
  }

  onCancel = (e) => {
    this.setVisible(false);

    const { onCancel } = this.props;
    if (onCancel) {
      onCancel.call(this, e);
    }
  }

  onVisibleChange = (visible) => {
    this.setVisible(visible);
  }

  setVisible(visible) {
    const props = this.props;
    if (!('visible' in props)) {
      this.setState({ visible });
    }

    const { onVisibleChange } = props;
    if (onVisibleChange) {
      onVisibleChange(visible);
    }
  }

  render() {
    const { prefixCls, title, placement, okText, cancelText, ...restProps } = this.props;
    const popconfirmLocale = this.getLocale();

    const overlay = (
      <div>
        <div className={`${prefixCls}-inner-content`}>
          <div className={`${prefixCls}-message`}>
            <Icon type="exclamation-circle" />
            <div className={`${prefixCls}-message-title`}>{title}</div>
          </div>
          <div className={`${prefixCls}-buttons`}>
            <Button onClick={this.onCancel} size="small">
              {cancelText || popconfirmLocale.cancelText}
            </Button>
            <Button onClick={this.onConfirm} type="primary" size="small">
              {okText || popconfirmLocale.okText}
            </Button>
          </div>
        </div>
      </div>
    );

    return (
      <Tooltip
        {...restProps}
        prefixCls={prefixCls}
        placement={placement}
        onVisibleChange={this.onVisibleChange}
        visible={this.state.visible}
        overlay={overlay}
      />
    );
  }
}

const injectPopconfirmLocale = injectLocale('Popconfirm', {
  cancelText: '取消',
  okText: '确定',
});
Popconfirm.defaultProps = {
  prefixCls: 'ant-popover',
  transitionName: 'zoom-big',
  placement: 'top',
  trigger: 'click',
};
Popconfirm.propTypes = {
  prefixCls: PropTypes.string,
  overlayClassName: PropTypes.string,
  placement:PropTypes.oneOf(['top','left','right','bottom','topLeft','topRight','bottomLeft','bottomRight','leftTop','leftBottom','rightTop','rightBottom']),
  builtinPlacements: PropTypes.object,
  visible:PropTypes.bool,
  onVisibleChange: PropTypes.func,
  mouseEnterDelay:PropTypes.number,
  mouseLeaveDelay:PropTypes.number,
  transitionName: PropTypes.string,
  trigger: PropTypes.oneOf([ 'hover','focus','click']),
  openClassName: PropTypes.string,
  arrowPointAtCenter:PropTypes.bool,
  // getTooltipContainer had been rename to getPopupContainer
  getTooltipContainer: PropTypes.func,
  getPopupContainer: PropTypes.func,
  children: PropTypes.element,
  title: PropTypes.node,
  onConfirm: PropTypes.func,
  onCancel: PropTypes.func,
  okText: PropTypes.node,
  cancelText: PropTypes.node,
};
Popconfirm.displayName = "Popconfirm";

export default injectPopconfirmLocale(Popconfirm);
