// @gag/popover-web/src/index.jsx
import '../style';
import React from 'react';
import PropTypes from 'prop-types';
import assign from 'object-assign';
import Tooltip from '@gag/tooltip-web';
import warning from '@gag/util-web/warning';


class Popover extends React.Component{
  refs: {
    tooltip: Tooltip,
  };

  getPopupDomNode() {
    return this.refs.tooltip.getPopupDomNode();
  }

  getOverlay() {
    const { title, prefixCls, content } = this.props;
    warning(
      !('overlay' in this.props),
      'Popover[overlay] is removed, please use Popover[content] instead, ' +
      'see: http://u.ant.design/popover-content',
    );
    return (
      <div>
        {title && <div className={`${prefixCls}-title`}>{title}</div>}
        <div className={`${prefixCls}-inner-content`}>
          {content}
        </div>
      </div>
    );
  }

  render() {
    const props = assign({}, this.props);
    delete props.title;
    return (
      <Tooltip
        {...props}
        ref="tooltip"
        overlay={this.getOverlay()}
      />
    );
  }
}
Popover.defaultProps = {
  prefixCls: 'ant-popover',
  placement: 'top',
  transitionName: 'zoom-big',
  trigger: 'hover',
  mouseEnterDelay: 0.1,
  mouseLeaveDelay: 0.1,
  overlayStyle: {},
};
Popover.propTypes = {
  prefixCls: PropTypes.string,
  overlayClassName: PropTypes.string,
  placement:PropTypes.oneOf(['top','left','right','bottom','topLeft','topRight','bottomLeft','bottomRight','leftTop','leftBottom','rightTop','rightBottom']),
  builtinPlacements: PropTypes.object,
  visible:PropTypes.bool,
  onVisibleChange:PropTypes.func,
  mouseEnterDelay:PropTypes.number,
  mouseLeaveDelay:PropTypes.number,
  transitionName: PropTypes.string,
  trigger: PropTypes.oneOf([ 'hover','focus','click']),
  openClassName: PropTypes.string,
  arrowPointAtCenter:PropTypes.bool,
  // getTooltipContainer had been rename to getPopupContainer
  getTooltipContainer:PropTypes.func,
  getPopupContainer:PropTypes.func,
  children:PropTypes.element,
  title:PropTypes.node,
  content:PropTypes.node,
};
Popover.displayName = "Popover";
module.exports=Popover;
