// @gag/dropdown-web/src/dropdown-button.jsx
import React from 'react';
import Button from '@gag/button-web';
import Icon from '@gag/icon-web';
import Dropdown from './dropdown';
import PropTypes from 'prop-types';
const ButtonGroup = Button.Group;
import classNames from 'classnames';

export default class DropdownButton extends React.Component {
  render() {
    const {
      type, overlay, trigger, align, children, className, onClick, prefixCls,
      disabled, visible, onVisibleChange, placement, ...restProps,
    } = this.props;
    const cls = classNames(prefixCls, className);

    const dropdownProps = {
      align,
      overlay,
      trigger: disabled ? [] : trigger,
      onVisibleChange,
      placement,
    };

    if ('visible' in this.props) {
      dropdownProps.visible = visible;
    }

    return (
      <ButtonGroup {...restProps} className={cls}>
        <Button type={type} onClick={onClick} disabled={disabled}>{children}</Button>
        <Dropdown {...dropdownProps}>
          <Button type={type} disabled={disabled}>
            <Icon type="down" />
          </Button>
        </Dropdown>
      </ButtonGroup>
    );
  }
}
DropdownButton.defaultProps = {
  placement: 'bottomRight',
  type: 'default',
  prefixCls: 'ant-dropdown-button',
};
DropdownButton.propTypes = {
  prefixCls:PropTypes.string,
  className:PropTypes.string,
  type:PropTypes.oneOf(['primary' , 'ghost' , 'dashed']),
  onClick:PropTypes.func,
  trigger:PropTypes.arrayOf(PropTypes.string),
  align:PropTypes.any,
  overlay:PropTypes.node,
  visible:PropTypes.bool,
  disabled:PropTypes.bool,
  onVisibleChange:PropTypes.func,
  style:React.CSSProperties,
  children:PropTypes.any,
  placement:PropTypes.oneOf(['topLeft','topCenter','topRight','bottomLeft','bottomCenter' , 'bottomRight']),
};
DropdownButton.displayName = "DropdownButton";
module.exports=DropdownButton;
