// @gag/input-web/src/Search.jsx
import React from 'react';
import classNames from 'classnames';
import Input from './Input';
import PropTypes from 'prop-types';
import Icon from '@gag/icon-web';

class Search extends React.Component {
  onSearch = () => {
    const { onSearch } = this.props;
    if (onSearch) {
      onSearch(this.input.refs.input.value);
    }
    this.input.refs.input.focus();
  }
  render() {
    const { className, prefixCls, ...others } = this.props;
    delete others.onSearch;
    const searchSuffix = (
      <Icon
        className={`${prefixCls}-icon`}
        onClick={this.onSearch}
        type="search"
      />
    );
    return (
      <Input
        className={classNames(prefixCls, className)}
        onPressEnter={this.onSearch}
        ref={node => this.input = node}
        suffix={searchSuffix}
        {...others}
      />
    );
  }
}
Search.defaultProps = {
  prefixCls: 'ant-input-search',
  onSearch() {},
};
Search.propTypes = {
  className: PropTypes.string,
  placeholder: PropTypes.string,
  prefixCls: PropTypes.string,
  defaultValue: PropTypes.any,
  value: PropTypes.any,
  onChange:PropTypes.func,
  onSearch:PropTypes.func,
  size:PropTypes.oneOf(['large','default','small']),
  disabled: PropTypes.bool,
  readOnly: PropTypes.bool,

};
Search.displayName = "Search";
module.exports=Search;
