// @gag/input-number-web/src/index.jsx
import React from 'react';
import '../style';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import RcInputNumber from 'rc-input-number';

class InputNumber extends React.Component {
  render() {
    const { className, size, ...others } = this.props;
    const inputNumberClass = classNames({
      [`${this.props.prefixCls}-lg`]: size === 'large',
      [`${this.props.prefixCls}-sm`]: size === 'small',
    }, className);

    return <RcInputNumber className={inputNumberClass} {...others} />;
  }
}
InputNumber.defaultProps = {
  prefixCls: 'ant-input-number',
  step: 1,
};
InputNumber.propTypes = {
  prefixCls: PropTypes.string,
  min: PropTypes.number,
  max: PropTypes.number,
  value: PropTypes.number,
  step: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
    ]),
  defaultValue: PropTypes.number,
  onChange:PropTypes.func,
  disabled: PropTypes.bool,
  size:PropTypes.oneOf([ 'large','small','default']),
  formatter:PropTypes.func,
  parser:PropTypes.func,
  placeholder: PropTypes.string,
  className: PropTypes.string,
};
InputNumber.displayName = "InputNumber";
module.exports=InputNumber;
