// @gag/pagination-web/src/Pagination.jsx
import React from 'react';
import RcPagination from 'rc-pagination';
import zhCN from 'rc-pagination/lib/locale/zh_CN';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import injectLocale from '@gag/locale-provider-web/injectLocale';
import Select from '@gag/select-web';
import MiniSelect from './MiniSelect';

class Pagination extends React.Component{
  render() {
    const { className, size, ...restProps } = this.props;
    const locale = this.getLocale();
    const isSmall = size === 'small';
    return (
      <RcPagination
        {...restProps}
        className={classNames(className, { mini: isSmall })}
        selectComponentClass={isSmall ? MiniSelect : Select}
        locale={locale}
      />
    );
  }
}
Pagination.defaultProps = {
  prefixCls: 'ant-pagination',
  selectPrefixCls: 'ant-select',
};
Pagination.propTypes = {
  total: PropTypes.number.isRequired,
  defaultCurrent: PropTypes.number,
  current: PropTypes.number,
  defaultPageSize: PropTypes.number,
  pageSize: PropTypes.number,
  onChange:PropTypes.func,
  showSizeChanger: PropTypes.bool,
  pageSizeOptions:PropTypes.arrayOf(PropTypes.string),
  onShowSizeChange:PropTypes.func,
  showQuickJumper: PropTypes.bool,
  showTotal:PropTypes.func,
  size: PropTypes.string,
  simple: PropTypes.bool,
  locale: PropTypes.object,
  className: PropTypes.string,
  prefixCls: PropTypes.string,
  selectPrefixCls: PropTypes.string,
};
Pagination.displayName = "Pagination";
const injectPaginationLocale = injectLocale('Pagination', zhCN);
export default injectPaginationLocale(Pagination);
