// @gag/locale-provider-web/src/index.jsx
import React from 'react';
import '../style';
import PropTypes from 'prop-types';
import { changeConfirmLocale } from '@gag/modal-web/src/locale';

class LocaleProvider extends React.Component{
  static propTypes = {
    locale: PropTypes.object,
  };

  static childContextTypes = {
    antLocale: PropTypes.object,
  };

  getChildContext() {
    return {
      antLocale: {
        ...this.props.locale,
        exist: true,
      },
    };
  }

  componentWillMount() {
    this.componentDidUpdate();
  }

  componentDidUpdate() {
    const { locale } = this.props;
    changeConfirmLocale(locale && locale.Modal);
  }

  componentWillUnMount() {
    changeConfirmLocale();
  }

  render() {
    return React.Children.only(this.props.children);
  }
}
LocaleProvider.propTypes = {
  locale:PropTypes.shape({
    Pagination: PropTypes.object,
    DatePicker: PropTypes.object,
    TimePicker:PropTypes.object,
    Calendar: PropTypes.object,
    Table: PropTypes.object,
    Modal: PropTypes.object,
    Popconfirm: PropTypes.object,
    Transfer: PropTypes.object,
    Select: PropTypes.object,
  }),
  children: PropTypes.element,
};
LocaleProvider.displayName = "LocaleProvider";
module.exports=LocaleProvider;
