// @gag/collapse-web/src/Collapse.jsx
import React from 'react';
import '../style';
import PropTypes from 'prop-types';
import RcCollapse from 'rc-collapse';
import classNames from 'classnames';
import animation from '@gag/util-web/openAnimation';

export class CollapsePanel extends React.Component {
}
CollapsePanel.propTypes={
  key: PropTypes.string.isRequired,
  header: PropTypes.node.isRequired,
  className: PropTypes.string,
}
class Collapse extends React.Component{
  static Panel: typeof CollapsePanel = RcCollapse.Panel;
  render() {
    const { prefixCls, className = '', bordered } = this.props;
    const collapseClassName = classNames({
      [`${prefixCls}-borderless`]: !bordered,
    }, className);
    return <RcCollapse {...this.props} className={collapseClassName} />;
  }
}

Collapse.defaultProps = {
  prefixCls: 'ant-collapse',
  bordered: true,
  openAnimation: { ...animation, appear() {} },
};
Collapse.propTypes = {
  activeKey:PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.string),
      PropTypes.string,
    ]),
  defaultActiveKey:PropTypes.arrayOf(PropTypes.string),
  /** 手风琴效果 */
  accordion:PropTypes.bool,
  onChange:PropTypes.func,
  className: PropTypes.string,
  bordered:PropTypes.bool,
  prefixCls: PropTypes.string,
};
Collapse.displayName = "Collapse";
module.exports=Collapse;
